File: Hard/2642. Design Graph With Shortest Path Calculator/2642. Design Graph With Shortest Path Calculator.js
/**
 * @param {number} n
 * @param {number[][]} edges
 */
var Graph = function(n, edges) {
    this.n = n;
    this.graph = new Map();

    for (const edge of edges) {
        this.addEdge(edge);
    }
};

/**
 * @param {number[]} edge
 * @return {void}
 */
Graph.prototype.addEdge = function(edge) {
    const [from, to, edgeCost] = edge;
    this.graph.set(from, (this.graph.get(from) || []).concat([[to, edgeCost]]));
};

/**
 * @param {number} node1
 * @param {number} node2
 * @return {number}
 */
Graph.prototype.shortestPath = function(node1, node2) {
    if (node1 === node2) return 0;

    const minTimes = new Array(this.n).fill(Infinity);
    minTimes[node1] = 0;

    const visited = new Set();

    while (visited.size !== this.n) {
        const [source, minTimeToSource] = this.getSourceWithMinTime(minTimes, visited);

        if (minTimeToSource === Infinity) {
            return minTimes[node2] === Infinity ? -1 : minTimes[node2];
        }

        visited.add(source);

        for (const [target, timeToTarget] of (this.graph.get(source) || [])) {
            if (visited.has(target)) {
                continue;
            }

            const newTimeToTarget = minTimeToSource + timeToTarget;
            const currentTimeToTarget = minTimes[target];

            if (newTimeToTarget < currentTimeToTarget) {
                minTimes[target] = newTimeToTarget;
            }
        }
    }

    return minTimes[node2] === Infinity ? -1 : minTimes[node2];
};

/**
 * @param {number[]} minTimes
 * @param {Set<number>} visited
 * @return {[number, number]}
 */
Graph.prototype.getSourceWithMinTime = function(minTimes, visited) {
    let source = -1;
    let minTime = Infinity;

    for (const [to, edgeCost] of minTimes.entries()) {
        if (visited.has(to)) {
            continue;
        }

        if (edgeCost < minTime) {
            minTime = edgeCost;
            source = to;
        }
    }

    return [source, minTime];
};

File: Hard/2642. Design Graph With Shortest Path Calculator/2642. Design Graph With Shortest Path Calculator.rs
use std::collections::{HashMap, BinaryHeap};
use std::cmp::Reverse;

struct Graph {
    nodes: i32,
    edges: HashMap<i32, Vec<(i32, i32)>>,
}

impl Graph {
    fn new(n: i32, edges: Vec<Vec<i32>>) -> Self {
        let mut graph = HashMap::new();
        for edge in edges {
            let (from, to, cost) = (edge[0], edge[1], edge[2]);
            graph.entry(from).or_insert(vec![]).push((to, cost));
        }
        Graph { nodes: n, edges: graph }
    }

    fn add_edge(&mut self, edge: Vec<i32>) {
        let (from, to, cost) = (edge[0], edge[1], edge[2]);
        self.edges.entry(from).or_insert(vec![]).push((to, cost));
    }

    fn shortest_path(&self, node1: i32, node2: i32) -> i32 {
        let mut distances = HashMap::new();
        let mut pq = BinaryHeap::new();

        distances.insert(node1, 0);
        pq.push(Reverse((0, node1)));

        while let Some(Reverse((cost, current))) = pq.pop() {
            if current == node2 {
                return cost;
            }

            if let Some(neighbors) = self.edges.get(&current) {
                for &(neighbor, edge_cost) in neighbors {
                    let new_cost = cost + edge_cost;
                    if !distances.contains_key(&neighbor) || new_cost < *distances.get(&neighbor).unwrap() {
                        distances.insert(neighbor, new_cost);
                        pq.push(Reverse((new_cost, neighbor)));
                    }
                }
            }
        }

        -1 // If no path is found
    }
}
